In [3]:
# установить свежую версию
!pip install scipy==1.5.2

     |████████████████████████████████| 25.9MB 167kB/s 
ERROR: tensorflow 2.3.0 has requirement scipy==1.4.1, but you'll have scipy 1.5.2 which is incompatible.
ERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.
  Found existing installation: scipy 1.4.1
    Uninstalling scipy-1.4.1:
      Successfully uninstalled scipy-1.4.1


In [2]:
import math
import numpy as np
import pandas as pd
from scipy import stats as st
import matplotlib.pyplot as plt 

#### Введение

Рассмотрим критерии, которые можно использовать для сравнения и выявления взаимосвязи качественных характеристик.

Такого рода задачи возникают в случае изучения предпочтения людей или выявления зависимости между лекарством и выживаемостью.

Предположим, что мы наблюдаем $n$ объектов

$O_1, ..., O_n$ 

Объектом может быть человек, некоторый предмет, например, автомобиль, может быть ответ человека.

Будем предполагать, что объект обладает двумя качественными признаками.

$A: A_1, ..., A_r$

$B: B_1, ..., B_s$

In [ ]:
pd.DataFrame({'B1': ['n11', 'n21', '...', 'nr1', 'n1'], 'B2': ['n12', 'n22', '...', 'nr2', 'n2'], '...': ['...', '...', '...', '...', '...'], 'Bs': ['n1s', 'n2s', '...', 'nrs', 'ns'], '': ['m1', 'm2', '...', 'mr', 'n'] }, index = ['A1', 'A2', '...', 'Ar', ''])

B1   B2  ...   Bs     
A1   n11  n12  ...  n1s   m1
A2   n21  n22  ...  n2s   m2
...  ...  ...  ...  ...  ...
Ar   nr1  nr2  ...  nrs   mr
      n1   n2  ...   ns    n

$m_i$ кол-во объектов обладающих значением $A_i$ признака $A$, независимо от того, какое значение принимает признак $B$.

$n_i$ кол-во объектов обладающих значением $B_i$ признака $B$, независимо от того, какое значение принимает признак $A$.

$n$ общее число рассматриваемых объектов.


Такого рода представление наблюдений для дальнейшего изучения взаимосвязи двух признаков, называется **таблица сопряжённости**.

#### Независимость признаков

Обозначим через $p_i$ вер-ть того, что признак $A$ у случайно взятого объекта, принял значение $A_i$.

$p_i = P(A_i), i = \overline {1, r}$

$q_i = P(B_j), j = \overline {1, s}$

$A$ и $B$ называются независимыми, если при любых $i$ и $j$:

$p_{ij} = P(A_i \cap B_i) = p_iq_j$

$p_{ij}$ вер-ть того, что случайно взятый объект имеет одновремено значение признака $A$ равное $A_i$ и признака $B$ равное $B_i$.

#### Гипотеза

$H_0: P(A_i \cap B_j) = p_iq_j$ для любых $i, j$

$H_1: \exists(i, j): P(A_i \cap B_j) \neq p_iq_j$ для любых $i, j$

#### Статистика

Воспользуемся для проверки гипотезы о независимости 2 признаков, критерием хи-квадрат. Будем сравнивать наблюдаемые частоты и ожидаемые.

$\chi^2 = \sum \limits_{i=1}^r \sum \limits_{j=1}^s \frac{(n_{ij} - np_iq_j)^2}{np_iq_j} $, где

$n_{ij}$ наблюдаемые частоты (эмпирические)

$np_iq_j$ ожидаемые частоты (теоритические)

Вычислить истинные значения ожидаемых частот мы не можем, т.к. мы не знаем какая ф-я распределения. Но мы можем найти оценки $p_i$ и $q_j$ с помощью метода максимального правдоподобия. Составим ф-ю правдоподобия 

$L= \frac{n!}{\prod\limits_{i = \overline {1,r} \\ j = \overline {1,s}} n_{ij!}} \prod\limits_{i = \overline {1,r} \\ j = \overline {1,s}} (p_i q_j)^{n_{ij}}$

$\sum \limits_{i=1}^r p_i = 1$ и $\sum \limits_{j=1}^s q_j = 1$

Точечные оценки

$\hat{p_i} = \frac{m_i}{n}, i = \overline {1,r}$

$m_i$ кол-во объектов обладающих значением $i$ из $r$, признаки расположенные по строкам.

$\hat{q_j} = \frac{n_j}{n}, j = \overline {1, s}$

$n_i$ кол-во объектов обладающих значением $i$ из $s$, признаки расположенные по столбцам.

#### Статистический критерий

Подставим в статистику, оценки вероятностей и получим статистический критерий

$\chi^2 = \sum \limits_{i=1}^r \sum \limits_{j=1}^s \frac{(n_{ij} - \frac{m_in_j}{n})^2}{\frac{m_in_j}{n}} \sim \chi^2_{(s-1) \cdot (r - 1)}$, где

$n_{ij}$ наблюдаемые частоты (эмпирические)

$\frac{m_in_j}{n}$ оценка веростности

$(s-1) \cdot (r - 1)$ произведение степеней свободы, где

$r$ кол-во признаков по строкам

$s$ кол-во признаков по столбцам

#### Scipy

На выходе первым идёт критерий, вторым p-value, третьим степени свободы, четвертым ожидаемые частоты, по которым проверяем условие $\frac{m_in_j}{n} \geq 5$

In [ ]:
obs = np.array([[10, 10, 20], [20, 20, 20]])
st.chi2_contingency(obs)

(2.7777777777777777, 0.24935220877729622, 2, array([[12., 12., 16.],
        [18., 18., 24.]]))

#### Критическая область

$V_k = (\chi^2_{1 - \alpha, (s-1) \cdot (r - 1)}, + \infty)$, где

$(s-1) \cdot (r - 1)$ произведение степеней свободы, где

$r$ кол-во признаков по строкам

$s$ кол-во признаков по столбцам

$\chi^2_в \in V_k \Rightarrow H_0$ отвергается.

#### Статистический критерий. Поправка Йейтса

В случае таблицы размерности 2х2, статистика критерия не редко принимает завышенные значения и следовательно, мы отвергаем нулевую гипотезу чаще, чем предполагали, т.е. не с вер-ю равную уровню значимости $\alpha$.

$\chi^2 = \sum \limits_{i=1}^r \sum \limits_{j=1}^s \frac{(|n_{ij} - \frac{m_in_j}{n}| - 0.5)^2}{\frac{m_in_j}{n}}$

#### Критическая область. Поправка Йейтса

$V_k = (\chi^2_{1 - \alpha, 1}, + \infty)$, где

$\chi^2_{1 - \alpha, 1}$ имеет 1 степень свободы

#### Важное условие!

Ожидаемые частоты для каждой клетки таблицы, должны быть больше 5.

$\frac{m_in_j}{n} \geq 5$

В случае, если в какой-то клетке данное условие не выполняется, нам нужно объединить соседние значения признака.

##### Задача №1

Утверждается, что результат действия лекарства зависит от способа его применения. Проверить это утверждение при α=0.05 по следующим данным:

In [ ]:
df = pd.DataFrame({'способы применения': ['A', 'B', 'C'], 'благополучный результат': [20, 23, 19], 'неблагополучный результат': [11, 17, 16]})
df

способы применения  благополучный результат  неблагополучный результат
0                  A                       20                         11
1                  B                       23                         17
2                  C                       19                         16

Найдите выборочное значение статистики критерия?

Найдите квантиль, участвующий в формировании критической области, и примите статистическое решение.

**Решение:**

Есть два признака, признак А (действие лекарства), признак B (способ применения). Утверждается, что они взаимосвязаны.

Нулевая гипотеза состоит в том, что признаки А и В зависимы, связаны друг с другом.

$H_0: \exists(i, j): P(A_i \cap B_j) \neq p_iq_j$ для любых $i, j$

$H_1: P(A_i \cap B_i) = p_iq_j $

$\chi^2 = \sum \limits_{i=1}^r \sum \limits_{j=1}^s \frac{(n_{ij} - \frac{m_in_j}{n})^2}{\frac{m_in_j}{n}} \sim \chi^2_{(s-1) \cdot (r - 1)}$

$(s-1) \cdot (r - 1)$ произведение степеней свободы, где

$r$ кол-во признаков по строкам

$s$ кол-во признаков по столбцам

In [ ]:
obs = np.array([df['благополучный результат'].values, df['неблагополучный результат'].values])
st.chi2_contingency(obs)

(0.7346258969958244,
 0.692592865228864,
 2,
 array([[18.13207547, 23.39622642, 20.47169811],
        [12.86792453, 16.60377358, 14.52830189]]))

Получили, что выборочное значение статистики критерия $\chi^2_в = 0.73$ и данная величина имеет 2 степени свободы, нам нужно будет вычислить квантиль с двумя степенями свободы $\chi^2_{1 - \alpha, 2}$

Найдите квантиль, участвующий в формировании критической области, и примите статистическое решение

$V_k = (\chi^2_{1 - \alpha, 2}, + \infty)$

In [ ]:
alpha = 0.05
st.chi2.ppf(1 - alpha, 2)

5.991464547107979

**Ответ:**

$\chi^2_в$ = 0.73

$V_k = (5.99, + \infty)$

$\chi^2_в \notin V_k \Rightarrow H_0$ не отвергается. Значит утверждение верное, результат действия лекарства зависит от приема.

##### Задача №2

У 150 мужчин и 150 женщин узнали, на что они обращают внимание при выборе автомобиля. Один из вопросов был: Важно ли при выборе автомобиля, сколько топлива потребляет автомобиль?

Результаты опроса следующие:

120 мужчин ответили, что очень важно,  28 - важно, 2 - не очень важно.

74 женщины ответили, что очень важно, 59 -  важно,  15 - не очень важно, 2 - не важно.

Зависит ли критерий выбора автомобиля от пола? Уровень значимости 0.05

Для проверки гипотезы воспользуйтесь критерием хи-квадрат.

Формат ответа: через запятую с пробелом

*  число степеней свободы функции распределения статистики критерия
*  значение статистики критерия
*  левая граница области принятия нулевой гипотезы
*  правая граница области принятия нулевой гипотезы
*  статистическое решение (1, если независимы, 0, если зависимы)

**Решение:**

Поставим гипотезу. Нам нужно узнать, есть ли зависимость между двумя признаками, признаками A (пол) и признаком B (ответ). Если они независимы, то должно выполнятся условие, при любых $i$ и $j$:

$p_{ij} = P(A_i \cap B_i) = p_iq_j$

$p_{ij}$ вер-ть того, что случайно взятый объект имеет одновремено значение признака $A$ равное $A_i$ и признака $B$ равное $B_i$.

Нулевая гипотеза состоит в том, что нет зависимости между признаками A и B.

$H_0: P(A_i \cap B_i) = p_iq_j $

$H_1: \exists(i, j): P(A_i \cap B_j) \neq p_iq_j$ для любых $i, j$

Составим таблицу с данными

In [7]:
pd.DataFrame({'': ['М', 'Ж'], 'очень важно': [120, 74], 'важно': [28, 59], 'не очень важно': [2, 15], 'не важно': [0, 2]})

очень важно  важно  не очень важно  не важно
0  М          120     28               2         0
1  Ж           74     59              15         2

In [9]:
df = pd.DataFrame({'': ['очень важно', 'важно', 'не очень важно', 'не важно'], 'М': [120, 28, 2, 0], 'Ж': [74, 59, 15, 2]})
df

М   Ж
0     очень важно  120  74
1           важно   28  59
2  не очень важно    2  15
3        не важно    0   2

Видим, что у последнего признака не выполняется условие $\frac{m_in_j}{n} \geq 5$, поэтому объединим два последних интервала.

In [24]:
n = df['М'].sum() + df['Ж'].sum()
m4 = 2
n2 = 150

m4 * n2 / n

1.0

In [25]:
df = pd.DataFrame({'': ['очень важно', 'важно', 'не очень важно'], 'М': [120, 28, 2], 'Ж': [74, 59, 17]})
df

М   Ж
0     очень важно  120  74
1           важно   28  59
2  не очень важно    2  17

In [30]:
df['М'].values

array([120,  28,   2])

число степеней свободы функции распределения статистики критерия. видим, третьим аргументов выхода функции, число степеней свободы 2.

выборочная статистика критерия, первый аргумент.

In [31]:
obs = np.array([df['Ж'].values, df['М'].values, ])
st.chi2_contingency(obs)

(33.79529876949751, 4.5861062807170865e-08, 2, array([[97. , 43.5,  9.5],
        [97. , 43.5,  9.5]]))

левая граница **области принятия** нулевой гипотезы

правая граница **области принятия** нулевой гипотезы

Критическая область

$V_k = (\chi^2_{1 - \alpha, (s-1) \cdot (r - 1)}, + \infty)$

$(s-1) \cdot (r - 1)$ произведение степеней свободы, где

$r$ кол-во признаков по строкам

$s$ кол-во признаков по столбцам

А нам нужно **область принятия**, т.к. наша статистика критерия имеет распределение $\chi^2$, в независимости от того, сколько степеней свободы имеет данное распределение, оно не принимает значения аргумента меньше нуля, по построению сл. величины $\chi^2$, которая $\chi^2 = \xi^2_1 + ... + \xi^2_k$, где $\xi_i \sim N(0, 1)$.

Поэтоме левая граница принятия решения будет 0. А правая граница принятия решения, будет квантиль $\chi^2_{1 - \alpha, 2}$ с 2 степенями свободы.

In [34]:
alpha = 0.05
st.chi2.ppf(1 - alpha, 2)

5.991464547107979

$V_k = (5.99, + \infty)$, где

$\chi^2_в = 33.79$

$\chi^2_в \in V_k \Rightarrow H_0$ отвергается. Значит признаки A и B зависимы.

**Ответ:**

1. Степеней свободы ф-и распределения статистики критерия 2.
2. 

#### Что делать, когда не выполнено условие $\frac{m_in_j}{n} \geq 5$. Точный критерий Фишера

Предположим, что есть $n$ объектов

$O_1, ..., O_n$

Каждый из объекта обладает двумя качественными признаками

$A: A_1, A_2$

$B: B_1, B_2$

In [ ]:
pd.DataFrame({'': ['A1', 'A2', 'сумма'], 'B1': ['n11', 'n21', 'n1'], 'B2': ['n12', 'n22', 'n2'], 'сумма': ['m1', 'm2', 'n']})

B1   B2 сумма
0     A1  n11  n12    m1
1     A2  n21  n22    m2
2  сумма   n1   n2     n

#### Гипотеза

Ответим на вопрос, зависимы ли признаки $A$ и $B$?

$H_0: A$ и $B$ зависимы

$H_1: A$ и $B$ независимы

Будем считать, что какой-то из $n_{ij} < 5$. В противном случае, мы бы могли воспользоваться критерий хи-квадрат, в данном случае хи-квадрат использовать мы не можем.

#### Точный критерий Фишера

Основан на сравнении вероятностей реализации таблиц с такими же суммарными значениями по строкам и столбцам, т.е. $n_j$ и $m_i$, что и исходная наблюдаемая таблица.

$T_0$ обозначим так исходную таблицу, так же добавим индекс с ноликом в таблицу. И зафиксируем $n_1, n_2, m_1, m_2$.

In [ ]:
pd.DataFrame({'': ['A1', 'A2', 'сумма'], 'B1': ['n_0_11', 'n_0_21', 'n1'], 'B2': ['n_0_12', 'n_0_22', 'n2'], 'сумма': ['m1', 'm2', 'n']})

B1      B2 сумма
0     A1  n_0_11  n_0_12    m1
1     A2  n_0_21  n_0_22    m2
2  сумма      n1      n2     n

При фиксированных $n_1, n_2, m_1, m_2$ рассмотрим все возможные заполнения таблицы. Обозначим одну из таких реализаций, через $T_i$, найдём вероятность получить таблицу $T_i$

In [ ]:
pd.DataFrame({'': ['A1', 'A2', 'сумма'], 'B1': ['n_i_11', 'n_i_21', 'n1'], 'B2': ['n_i_12', 'n_i_22', 'n2'], 'сумма': ['m1', 'm2', 'n']})

$P_i = \frac{C^{n^i_{11}, n^i_{12}, n^i_{21}, n^i_{22}}_n}{C^{n_1, n_2}_n \cdot C^{m_1, m_2}_n} = \frac{\cdot n_{1}! \cdot n_{2}! \cdot m_{1}! \cdot m_{2}!}{n^i_{11}! \cdot n^i_{12}! \cdot n^i_{21}! \cdot n^i_{22}! \cdot n! }$

#### Статистика критерия Фишера 

Обозначим через $P_0$ вероятность реализации таблицы $T_0$ (исходной таблицы по нашим наблюдениям).



Статистика двухстороннего критерия

$P = \sum \limits_{T_i: P_i \leq P_0, i \neq 0} P_i + P_0$

это вероятность равная сумме вероятностей, всех таблиц $T_i$ для которых вероятность реализации, не превосходит заданной вероятности $P_0$.

Если $P < \alpha \Rightarrow H_0$ отклоняем. Значит признаки зависимы!

#### Scipy

Подаём в массив значения из изначальной таблицы. pvalue и есть статистика критерия $P = \sum \limits_{T_i: P_i \leq P_0, i \neq 0} P_i + P_0$

In [ ]:
oddsratio, pvalue = st.fisher_exact([[8, 2], [1, 5]])
oddsratio, pvalue

(20.0, 0.03496503496503495)

#### Пример

Два эксперта должны были разметить одни и теже 10 фотографий с улыбающимися людьми. Если они считали, что человек на фото улыбается естественно, ставили 1, иначе 0.

In [ ]:
pd.DataFrame({'№ улыбки': ['exp1', 'exp2'], '1': [1, 1], '2': [1, 1], '3': [0, 1], '4': [1, 0], '5': [1, 0], '6': [0, 0], '7': [0, 1], '8': [1, 1], '9': [1, 1], '10': [0, 0]})

№ улыбки  1  2  3  4  5  6  7  8  9  10
0     exp1  1  1  0  1  1  0  0  1  1   0
1     exp2  1  1  1  0  0  0  1  1  1   0

Проверить гипотезу о том, что вероятность поставить 1 одна и та же у двух экспертов, т.е. оценки экспертов однородны.

In [ ]:
pd.DataFrame({'': ['exp1', 'exp2', 'сумма'], '1': [6, 6, 12], '0': [4, 4, 8], 'сумма': [10, 10, 20]})

1  0  сумма
0   exp1   6  4     10
1   exp2   6  4     10
2  сумма  12  8     20

Видим, что одинаковое количество 1 и 0 поставили оба эксперта, поэтому проверять гипотезу об однородности мы не будем, будем считать, что вероятность поставить 1 для экспертов равны.

Мы можем проверить гипотезу о том, есть ли взаимосвязь между оценками 1 и 2 эксперта. Была ли например, какая-нибудь инструкция, которой оба эксперта следовали.

Составим таблицу

In [ ]:
pd.DataFrame({'exp1/exp2': [0, 1, 'сумма'], '0': [2, 2, 4], '1': [2, 4, 6], 'сумма': [4, 6, 10]})

exp1/exp2  0  1  сумма
0         0  2  2      4
1         1  2  4      6
2     сумма  4  6     10

На пересечении строки 0 и столбца 0, стоит кол-во раз, когда оба эксперта поставили фотографии по нулю.

На пересечении строки 1 и столбца 0, стоит кол-во раз, когда первый эксперт поставил фотографии 1, а второй поставил 0. 

Проверим гипотезу о независимости оценок первого эксперта от оценок второго эксперта, с помощью критерия Фишера.

Для этого нам нужно перебрать все возможные таблицы, такие, что в сумме они дают по строкам значения 4, 6 и по столбцам 4, 6. И вычислить вероятность таких таблиц. 

Обозначим данную таблицу $T_0$ и вычислим её вероятность.

$P_i = \frac{C^{n^i_{11}, n^i_{12}, n^i_{21}, n^i_{22}}_n}{C^{n_1, n_2}_n \cdot C^{m_1, m_2}_n} = \frac{\cdot n_{1}! \cdot n_{2}! \cdot m_{1}! \cdot m_{2}!}{n^i_{11}! \cdot n^i_{12}! \cdot n^i_{21}! \cdot n^i_{22}! \cdot n! }$

$P_0 = \frac{4!6!4!6!}{10!2!2!2!4!} = 0.429$

Для того, чтобы построить другие таблицы, воспользуемся элементом из строки 0, столбца 0, и начнём менять его от 0 до 4, пропустив 2.

In [ ]:
# T_1
pd.DataFrame({'exp1/exp2': [0, 1, 'сумма'], '0': [0, 4, 4], '1': [4, 2, 6], 'сумма': [4, 6, 10]})

exp1/exp2  0  1  сумма
0         0  0  4      4
1         1  4  2      6
2     сумма  4  6     10

$T_1$

$P_1 = 0.071$

In [ ]:
# T_2
pd.DataFrame({'exp1/exp2': [0, 1, 'сумма'], '0': [1, 3, 4], '1': [3, 3, 6], 'сумма': [4, 6, 10]})

$T_2$

$P_2 = 0.071$

In [ ]:
# T_3
pd.DataFrame({'exp1/exp2': [0, 1, 'сумма'], '0': [3, 1, 4], '1': [1, 5, 6], 'сумма': [4, 6, 10]})

$T_3$

$P_3 = 0.114$

In [ ]:
# T_4
pd.DataFrame({'exp1/exp2': [0, 1, 'сумма'], '0': [4, 0, 4], '1': [0, 6, 6], 'сумма': [4, 6, 10]})

$T_4$

$P_4 = 0.005$

Используем статистику

$P = \sum \limits_{T_i: P_i \leq P_0, i \neq 0} P_i + P_0$



Все таблицы построенные на основе исходной, имеют вероятность меньшую, чем вероятность $P_0$, поэтому при расчёте значения статистики критерия $P = 1$, мы должны просуммировать все вероятности

$P = P_0 + P_1 + P_2 + P_3 + P_4 = 1$

$P > \alpha \Rightarrow H_0$ не отвергается. Оценки 1 и 2 эксперта независимы.

**Ответ:**

$H_0$ отвергается. Значит оценки экспертов независимы.

#### Задача №1

Можно ли снизить риск тромбоза назначением небольших доз аспирина. Было проведено контролируемое испытание. Все больные, согласившиеся на участие в испытании и не имевшие противопоказании к аспирину, были случайным образом разделены на две группы: 1-я получала плацебо, 2-я — аспирин.

B 1-й группе тромбоз произошел у 3 из 4 больных, во 2-й — у 1 из 3. Можно ли говорить о статистически значимом 
различии доли больных с тромбозом, а тем самым об эффективности аспирина?

Воспользуйтесь критерием Фишера для проверки гипотезы. Уровень значимости 0.05

Чему равна вероятность реализации исходной таблицы?

Найдите статистику критерия и примите статистическое решение

**Решение:**

1 - произошёл тромбоз

0 - тромбоза не произошло

$H_0: A$ и $B$ доли одинаковы

$H_0: A$ и $B$ доли различны

In [ ]:
# T_0
pd.DataFrame({'group_1/group_2': [0, 1, 'sum'], '0': [1, 3, 4], '1': [2, 1, 3], 'sum': [3, 4, 7]})

group_1/group_2  0  1  sum
0               0  1  2    3
1               1  3  1    4
2             sum  4  3    7

Решим с использованием Scipy. Подадим в функцию значения исходной таблицы, pvalue и есть статистика критерия $P = \sum \limits_{T_i: P_i \leq P_0, i \neq 0} P_i + P_0$

In [ ]:
oddsratio, pvalue = st.fisher_exact([[1, 2], [3, 1]])
oddsratio, pvalue

(0.16666666666666666, 0.4857142857142858)

Найдём вер-ть $P_0$ реализации данной таблицы $T_0$



$P_i = \frac{\cdot n_{1}! \cdot n_{2}! \cdot m_{1}! \cdot m_{2}!}{n^i_{11}! \cdot n^i_{12}! \cdot n^i_{21}! \cdot n^i_{22}! \cdot n! }$

$P_0 = \frac{4!3!3!4!}{1!2!3!1!7!}$

In [ ]:
def count_prob(a, b):
  s = 1
  r = 1
  for x in a:
    s *= math.factorial(x)

  for x in b:
    r *= math.factorial(x)

  return s/r

In [ ]:
a = [4, 3, 3, 4]
b = [1, 2, 3, 1, 7]

p0 = count_prob(a, b)
p0

0.34285714285714286

Найдите статистику критерия и примите статистическое решение

Зафиксируем суммы значений 3, 4. Переберем все возможные таблицы, которые дают в сумме 3,4. Возьмем элемент из 0 строки и 0 столбца и будем его менять.

In [ ]:
# T_1
pd.DataFrame({'group_1/group_2': [0, 1, 'sum'], '0': [0, 4, 4], '1': [3, 0, 3], 'sum': [3, 4, 7]})

group_1/group_2  0  1  sum
0               0  0  3    3
1               1  4  0    4
2             sum  4  3    7

In [ ]:
a = [4, 3, 3, 4]
b = [0, 3, 4, 0, 7]

p1 = count_prob(a, b)
p1

0.02857142857142857

In [ ]:
# T_2
pd.DataFrame({'group_1/group_2': [0, 1, 'sum'], '0': [2, 2, 4], '1': [1, 2, 3], 'sum': [3, 4, 7]})

group_1/group_2  0  1  sum
0               0  2  1    3
1               1  2  2    4
2             sum  4  3    7

In [ ]:
a = [4, 3, 3, 4]
b = [2, 1, 2, 2, 7]

p2 = count_prob(a, b)
p2

0.5142857142857142

In [ ]:
# T_3
pd.DataFrame({'group_1/group_2': [0, 1, 'sum'], '0': [3, 1, 4], '1': [0, 3, 3], 'sum': [3, 4, 7]})

group_1/group_2  0  1  sum
0               0  3  0    3
1               1  1  3    4
2             sum  4  3    7

In [ ]:
a = [4, 3, 3, 4]
b = [3, 0, 1, 3, 7]

p3 = count_prob(a, b)
p3

0.11428571428571428

Вычислим статистику критерия

$P = \sum \limits_{T_i: P_i \leq P_0, i \neq 0} P_i + P_0$

нужно просуммировать все такие вероятности $P_i$, которые $P_i \leq P_0$

In [ ]:
np.sum([pi for pi in [p0, p1, p2, p3] if pi <= p0])

0.4857142857142857

$\alpha$ = 0.05

$P > \alpha \Rightarrow H_0$ не отклоняется. Доли одинаковы.

**Ответ:**

$H_0$ не отклоняется. Доли одинаковы.

#### Код по теме

In [ ]:
import scipy as sc
import math as mt
from scipy import stats as st
import matplotlib.pyplot as plt

# Исходная таблица сопряженности
k11, k12, k21, k22 = 2,4,3,5 # 2,1,1,3 # 2,4,3,5 # 2,1,1,3 # 6,4,6,4
kk1 = sc.array([[k11, k12], [k21, k22]])

kk = sortmin(kk1) # сортируем таблицу сопряженности

n = sc.array([kk[k,:].sum()  for k in range(kk.shape[0])])
m = sc.array([kk[:,k].sum()  for k in range(kk.shape[1])])
nm = sc.array([n,m])

# вычисление вероятностей ps для всех возможных реализаций таблиц Ti:

vfact = sc.vectorize(lambda n: float(mt.factorial(n))) # векторизованный факториал
pf = vfact(nm).prod()/float(mt.factorial(kk.sum()))
ps = sc.array([pf/vfact(get_tab(x, n, m)).prod() for x in range(min(m[0], n[0])+1)]) # min(m[0], n[0])

# вычисление СТАТИСТИК p_less,p_greater, p_two_sided: 

kmin = kk[0,0]
p_less = ps[:kmin+1].sum() 
p_greater = ps[kmin:].sum()
p_two_sided = p_less + ps[kmin+1: ][ps[kmin+1:] <= ps[kmin] ].sum() 

# проверка СТАТИСТИК p_less,p_greater, p_two_sided с использованием 
# scipy.statistic.fisher_exact(...):

test_p_less = st.fisher_exact(kk1, alternative='less')[1]
test_p_greater = st.fisher_exact(kk, alternative='greater')[1]
test_p_two_sided = st.fisher_exact(kk, alternative='two-sided')[1]

# Рисуем график распределения вероятностей p(Ti):
plt.grid()
plt.plot(range(len(ps)), ps, 'o', label='ps')
plt.vlines(range(len(ps)),0, ps)
plt.legend()
plt.show()

#===================
def sortmin(arr):
    argmin = arr.argmin()
    row_num = argmin//arr.shape[1]
    col_num = argmin%arr.shape[1]
    
    col_0 = arr[:,0].copy()
    arr[:,0] = arr[:,col_num]
    arr[:,col_num] = col_0
    
    row_0 = arr[0, :].copy()
    arr[0, :] =  arr[row_num, :]
    arr[row_num, :] = row_0
    
    return arr
#=============================
def get_tab(x, n, m):
    return sc.array([[x, n[0]-x],[m[0]-x, n[1]-m[0]+x]])
#=============================

Код, как можно реализовать статистический критерий $\chi^2 = \sum \limits_{i=1}^r \sum \limits_{j=1}^s \frac{(n_{ij} - \frac{m_in_j}{n})^2}{\frac{m_in_j}{n}}$

In [ ]:
import numpy as np
from scipy import stats as st

nij = np.array([...]) # Таблица сопряженности
n= nij.sum() # общее кол-во наблюдений в таблице сопряженности

mi = np.array([[sum(nij[i,:]) for i in range(nij.shape[0])]]).T
nj = np.array([[sum(nij[:, j]) for j in range(nij.shape[1])]])

minj = np.dot(mi, nj) # таблица (матрица) произведений mi*nj

# Формула расчета выборочного Хи-квадрат:

xi2v = ((nij- minj/n)**2 / (minj/n)).sum()

# тестовый расчет для проверки формулы 
# с использованием библиотечной функции
# scipy.stats.chi2_contingency

xi2_test = st.chi2_contingency(nij)[0]

print('xi2v:', xi2v)
print('xi2_test:', xi2_test)